pkg/autoscaler/collector.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2019 The Knative Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package autoscaler
+
+import (
+	"context"
+	"sync"
+
+	kpa "github.com/knative/serving/pkg/apis/autoscaling/v1alpha1"
+	"go.uber.org/zap"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Metric represents a resource to configure the metric collector with.
+// +k8s:deepcopy-gen=true
+type Metric struct {
+	metav1.ObjectMeta
+	Spec   MetricSpec
+	Status MetricStatus
+}
+
+// MetricSpec contains all values the metric collector needs to operate.
+type MetricSpec struct{}
+
+// MetricStatus reflects the status of metric collection for this specific entity.
+type MetricStatus struct{}
+
+// NewMetricCollector creates a new metric collector.
+func NewMetricCollector(logger *zap.SugaredLogger) *MetricCollector {
+	collector := &MetricCollector{
+		logger:      logger,
+		collections: make(map[string]*collection),
+	}
+
+	return collector
+}
+
+// MetricCollector manages collection of metrics for many entities.
+type MetricCollector struct {
+	logger *zap.SugaredLogger
+
+	statsScraperFactory StatsScraperFactory
+
+	metrics map[string]*Metric
+
+	collections      map[string]*collection
+	collectionsMutex sync.RWMutex
+}
+
+// Get gets a Metric's state from the collector.
```

Comment:
It deserves being noted that a copy is returned, i.e. it's mutable on its own (or that changes to it won't be seen in the collector).

Reply:
Done